banditpylib/learners/collaborative_learner/collaborative_learning.py

```diff
@@ -0,0 +1,390 @@
+
+  @property
+  def best_arm(self) -> int:
+    # returns arm that the agent chose (could be None)
+    if self.__stage != "termination":
+      raise Exception('%s: I don\'t have an answer yet!' % self.name)
+    return self.__learning_arm
+
+  def broadcast(self) -> Dict[int, Tuple[float, int]]:
+    if self.__stage != "communication":
+      raise Exception('%s: I can\'t broadcast in stage %s!'\
+        % (self.name, self.__stage))
+    return_dict = {}
+    if self.__learning_arm:
+      return_dict[self.__learning_arm] = (self.__learning_mean,
+        self.__pulls_used)
+    self.__complete_round()
+    return return_dict
+
+class LilUCBHeuristicCollaborativeBAIMaster(CollaborativeBAIMaster):
+  r"""Implementation of master in Collaborative Learning Algorithm
+
+  :param int arm_num: number of arms of the bandit
+  :param int rounds: number of rounds of communication allowed
+    (this agent uses one more)
+  :param int horizon: maximum number of pulls the agent can make
+    (over all rounds combined)
+  :param int num_agents: number of agents
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, arm_num:int, rounds: int,
+    horizon: int, num_agents: int, name: Optional[str] = None):
+    super().__init__(name)
+    if arm_num <= 1:
+      raise ValueError('Number of arms is expected at least 2. Got %d.' %
+                       arm_num)
+    if rounds <= 1:
+      raise ValueError('Number of rounds is expected at least 2. Got %d.' %
+                       rounds)
+    self.__arm_num = arm_num
+    self.__comm_rounds = rounds - 1
+    self.__T = horizon
+    self.__num_agents = num_agents
+
+  def _name(self) -> str:
+    return "lilUCBHeur_collaborative_master"
+
+  def reset(self):
+    self.__active_arms = list(range(self.__arm_num))
+
+  def __assign_arms(self, agent_ids: List[int]) ->\
+    Dict[int, List[int]]:
+    agent_arm_assignment: Dict[int, List[int]] = {}
+
+    def random_round(x: float) -> int:
+      # if x = 19.3
+      # rounded to 19 with probability 0.7
+      # rounded to 20 with probability 0.3
+      frac_x = x - int(x)
+      if np.random.uniform() > frac_x:
+        return int(x)
+      return int(x) + 1
+
+    if len(self.__active_arms) <= 1:
+      # don't assign any arms if 0/1 active arms
+      agent_arm_assignment = {}
+      for agent_id in agent_ids:
+        agent_arm_assignment[agent_id] = []
+      return agent_arm_assignment
+
+    if len(self.__active_arms) < len(agent_ids):
+      arms_assign_list = []
+      num_agents_per_arm = len(agent_ids) / len(self.__active_arms)
+      for arm in self.__active_arms[:-1]:
+        arms_assign_list += [[arm]] * \
+          random_round(num_agents_per_arm)
+      arms_assign_list += [[self.__active_arms[-1]]] * \
+        (len(agent_ids) - len(arms_assign_list))
+      random.shuffle(arms_assign_list)
+      for i, agent_id in enumerate(agent_ids):
+        agent_arm_assignment[agent_id] = arms_assign_list[i]
```

Comment:
You could use `numpy.random.multinomial` to assign arms.

Reply:
I thought of using something similar too, but this might lead to randomness.
Eg. we might want to assign 4 arms to each agent, but end up assigning 2 to one and 6 to another.

Reply:
Is it okay to if the number of arms assigned is random, but with the expectation same as `num_arms/num_agents`

Reply:
Okay, I think your method seems more robust. However, I think either way is fine. You can remain to use this method.